core/src/main/scala/org/apache/spark/sql/delta/commands/merge/MergeIntoMaterializeSource.scala

```diff
@@ -0,0 +1,439 @@
+import org.apache.spark.sql.delta.files.TahoeFileIndex
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+import org.apache.spark.SparkException
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{AttributeSet, Expression, ExpressionSet, Literal}
+import org.apache.spark.sql.catalyst.optimizer.{EliminateResolvedHint, JoinSelectionHelper}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.execution.{LogicalRDD, SQLExecution}
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Trait with logic and utilities used for materializing a snapshot of MERGE source
+ * in case we can't guarantee deterministic repeated reads from it.
+ *
+ * We materialize source if it is not safe to assume that it's deterministic
+ * (override with MERGE_SOURCE_MATERIALIZATION).
+ * Otherwise, if source changes between the phases of the MERGE, it can produce wrong results.
+ * We use local checkpointing for the materialization, which saves the source as a
+ * materialized RDD[InternalRow] on the executor local disks.
+ *
+ * 1st concern is that if an executor is lost, this data can be lost.
+ * However, when Spark executor decomissioning API is used, it should attempt to move this
+ * materialized data safely out before removing the executor.
+ *
+ * 2nd concern is that if an executor is lost for another reason (e.g. spot kill), we will
+ * still lose that data. To mitigate that, we implement a retry loop.
+ * The whole Merge operation needs to be restarted from the beginning in this case.
+ * When we retry, we increase the replication level of the materialized data from 1 to 2.
+ * (override with MERGE_SOURCE_MATERIALIZATION_RDD_STORAGE_LEVEL_RETRY).
+ * If it still fails after the maximum number of attempts (MERGE_MATERIALIZE_SOURCE_MAX_ATTEMPTS),
+ * we record the failure for tracking purposes.
+ *
+ * 3rd concern is that executors run out of disk space with the extra materialization.
+ * We record such failures for tracking purpuses.
+ */
+trait MergeIntoMaterializeSource extends DeltaLogging {
+  import MergeIntoMaterializeSource._
+
+  /**
+   * Prepared Dataframe with source data.
+   * If needed, it is materialized, @see prepareSourceDF
+   */
+  protected var sourcePlanDF: Dataset[Row] = null
+
+  /**
+   * If the source was materialized, reference to the checkpointed RDD.
+   */
+  protected var materializedSourceRDD: Option[RDD[InternalRow]] = None
+
+  /**
+   * Track which attempt or retry it is in runWithMaterializedSourceAndRetries
+   */
+  protected var attempt: Int = 0
+
+  /**
+   * This will run the Merge with retries in case it detects an RDD block lost error of the
```

Comment:
```suggestion
   * Run the Merge with retries in case it detects an RDD block lost error of the
```
Good method docs follow either "verb" first approach. See Spark Dataframe docs - https://spark.apache.org/docs/1.6.3/api/java/org/apache/spark/sql/DataFrame.html#filter(org.apache.spark.sql.Column)